src/sentry/features/base.py

```diff
@@ -1,24 +1,40 @@
 __all__ = ["Feature", "OrganizationFeature", "ProjectFeature", "ProjectPluginFeature"]
 
+from abc import ABC
+from typing import TYPE_CHECKING, Any
 
-class Feature:
-    def __init__(self, name):
+if TYPE_CHECKING:
+    from sentry.models import Organization, Project
+
+
+class Feature(ABC):
+    def __init__(self, name: str, *args: Any, **kwargs: Any) -> None:
+        """
+        `FeatureManager.get()` and `FeatureCheckBatch.get_feature_objects()`
+        expect to be able to pass a `Feature` arbitrary `args` and `kwargs`.
+        """
         self.name = name
 
+    def get_organization(self) -> "Organization":
+        raise NotImplementedError
+
 
 class OrganizationFeature(Feature):
-    def __init__(self, name, organization):
-        Feature.__init__(self, name)
+    def __init__(self, name: str, organization: "Organization") -> None:
+        super().__init__(name)
         self.organization = organization
 
+    def get_organization(self) -> "Organization":
+        return self.organization
+
 
-class ProjectFeature(Feature):
-    def __init__(self, name, project):
-        Feature.__init__(self, name)
+class ProjectFeature(OrganizationFeature):
+    def __init__(self, name: str, project: "Project") -> None:
+        super().__init__(name, organization=project.organization)
```

Comment:
`ProjectFeature` now inherits from OrganizationFeature, and accesses `organization` while previously it didn't. That's an additional unconditional DB query where previously it may have been conditional and only in some handlers.

I would definetly recommend against runtime changes within the same PR that adds typing. Instead of changing the type hierarchy I would implement `get_organization()` for ProjectFeature explicitly.